eth_trie/types.nim

```diff
@@ -20,12 +34,128 @@ type
   DelProc = proc (db: RootRef, key: openarray[byte])
   ContainsProc = proc (db: RootRef, key: openarray[byte]): bool
 
-  TrieDatabaseRef* = object
+  TrieDatabaseRef* = ref object
     obj: RootRef
     putProc: PutProc
     getProc: GetProc
     delProc: DelProc
     containsProc: ContainsProc
+    mostInnerTransaction: Transaction
+
+  Transaction* = ref object
+    db: TrieDatabaseRef
+    parentTransaction: Transaction
+    modifications: MemDB
+    committed: bool
+
+proc put*(db: TrieDatabaseRef, key, val: openarray[byte])
+proc get*(db: TrieDatabaseRef, key: openarray[byte]): Bytes
+proc del*(db: TrieDatabaseRef, key: openarray[byte])
+proc beginTransaction*(db: TrieDatabaseRef): Transaction
+
+# TODO: This should be commited upstream
+proc `==` *[T](x, y: openarray[T]): bool =
+  if x.len != y.len:
+    return false
+
+  for f in low(x)..high(x):
+    if x[f] != y[f]:
+      return false
+
+  result = true
+
+proc keccak*(r: BytesRange): KeccakHash =
+  keccak256.digest r.toOpenArray
+
+let
+  # XXX: turning this into a constant leads to a compilation failure
+  emptyRlp = rlp.encode ""
+  emptyRlpHash = emptyRlp.keccak
+
+proc get*(db: MemDB, key: openarray[byte]): Bytes =
+  result = db.records.getOrDefault(@key).value
+  traceGet key, result
+
+proc del*(db: MemDB, key: openarray[byte]) =
+  traceDel key
+
+  # The database should ensure that the empty key is always active:
+  if key != emptyRlpHash.data:
+    # TODO: This is quite inefficient and it won't be necessary once
+    # https://github.com/nim-lang/Nim/issues/7457 is developed.
+    let key = @key
+
+    db.records.withValue(key, v):
+      dec v.refCount
+      if v.refCount <= 0:
+        db.records.del(key)
+        db.deleted.incl(key)
+
+proc contains*(db: MemDB, key: openarray[byte]): bool =
+  db.records.hasKey(@key)
+
+proc put*(db: MemDB, key, val: openarray[byte]) =
+  tracePut key, val
+
+  # TODO: This is quite inefficient and it won't be necessary once
+  # https://github.com/nim-lang/Nim/issues/7457 is developed.
+  let key = @key
+
+  db.deleted.excl(key)
+
+  db.records.withValue(key, v) do:
+    inc v.refCount
+  do:
+    db.records[key] = MemDBRec(refCount: 1, value: @val)
+
+proc newMemDB*: MemDB =
+  result.new
+  result.records = initTable[Bytes, MemDBRec]()
+  result.deleted = initSet[Bytes]()
+
+proc commit(memDb: MemDB, db: TrieDatabaseRef) =
+  for k in memDb.deleted:
+    db.del(k)
+
+  for k, v in memDb.records:
+    db.put(k, v.value)
+
+proc init(db: var MemDB) =
+  db = newMemDB()
+
+proc newMemoryDB*: TrieDatabaseRef =
+  new result
```

Comment:
This is a bit tricky. The new way to create in-memory databases is to just start a transaction that will never be rolled back or committed. We'll be able to use the same trick for the transient databases required for stateless clients.